### ISSUE TITLE
ARGB&I420ARemoteVideoFrameReady not working

### ISSUE THREAD

**Issue by dsp85103** [microsoft/MixedReality-WebRTC#121]
I using MixedReality-WebRTC in my case. It's WindowsPC to Android.
Android application is using library org.webrtc:google-webrtc:1.0.24180.
And peer connection is successful. It seem like this....
https://imgur.com/a/0yevhNd

Android could render frame from remote (windows pc). 
But PC side couldn't render frame from android. It seem like TrackAdded and I420ARemoteVideoFrameReady and ARGBRemoteVideoFrameReady not invoke by peer connection.. I dont know how to debug this .. Can you help me solve this problem?

Thanks

**Comment by djee-ms** [comment]
If `TrackAdded` is not called, it means the connection was not established correctly. Can you paste the SDP answer and offer messages? What signaling solution are you using? MixedReality-WebRTC only ships with NodeDssSignaler for C#; did you implement a version for Android?

**Comment by dsp85103** [comment]
Sorry~ Thank for ur reply. Here's offer and answer.

Offer from Android:
v=0
o=- 8015088664284850015 2 IN IP4 127.0.0.1
s=-
t=0 0
a=group:BUNDLE audio video data
a=msid-semantic: WMS ARDAMS
m=audio 9 UDP/TLS/RTP/SAVPF 111 103 104 9 102 0 8 106 105 13 110 112 113 126
c=IN IP4 0.0.0.0
a=rtcp:9 IN IP4 0.0.0.0
a=ice-ufrag:Len2
a=ice-pwd:w0zCZt7ROLAngg0y/eFjD4Yj
a=ice-options:trickle renomination
a=fingerprint:sha-256 D9:45:AC:F2:1A:2D:2A:25:10:64:E1:28:D3:C9:33:D1:47:41:53:54:02:44:D2:45:4D:32:83:5E:AD:3E:AC:0D
a=setup:actpass
a=mid:audio
a=extmap:1 urn:ietf:params:rtp-hdrext:ssrc-audio-level
a=sendrecv
a=rtcp-mux
a=rtpmap:111 opus/48000/2
a=rtcp-fb:111 transport-cc
a=fmtp:111 minptime=10;useinbandfec=1
a=rtpmap:103 ISAC/16000
a=rtpmap:104 ISAC/32000
a=rtpmap:9 G722/8000
a=rtpmap:102 ILBC/8000
a=rtpmap:0 PCMU/8000
a=rtpmap:8 PCMA/8000
a=rtpmap:106 CN/32000
a=rtpmap:105 CN/16000
a=rtpmap:13 CN/8000
a=rtpmap:110 telephone-event/48000
a=rtpmap:112 telephone-event/32000
a=rtpmap:113 telephone-event/16000
a=rtpmap:126 telephone-event/8000
a=ssrc:985434751 cname:duTyMEGqb0G2cnOB
a=ssrc:985434751 msid:ARDAMS 102
a=ssrc:985434751 mslabel:ARDAMS
a=ssrc:985434751 label:102
m=video 9 UDP/TLS/RTP/SAVPF 96 98 100 127 125 97 99 101 124
c=IN IP4 0.0.0.0
a=rtcp:9 IN IP4 0.0.0.0
a=ice-ufrag:Len2
a=ice-pwd:w0zCZt7ROLAngg0y/eFjD4Yj
a=ice-options:trickle renomination
a=fingerprint:sha-256 D9:45:AC:F2:1A:2D:2A:25:10:64:E1:28:D3:C9:33:D1:47:41:53:54:02:44:D2:45:4D:32:83:5E:AD:3E:AC:0D
a=setup:actpass
a=mid:video
a=extmap:2 urn:ietf:params:rtp-hdrext:toffset
a=extmap:3 http://www.webrtc.org/experiments/rtp-hdrext/abs-send-time
a=extmap:4 urn:3gpp:video-orientation
a=extmap:5 http://www.ietf.org/id/draft-holmer-rmcat-transport-wide-cc-extensions-01
a=extmap:6 http://www.webrtc.org/experiments/rtp-hdrext/playout-delay
a=sendrecv
a=rtcp-mux
a=rtcp-rsize
a=rtpmap:96 VP8/90000
a=rtcp-fb:96 ccm fir
a=rtcp-fb:96 nack
a=rtcp-fb:96 nack pli
a=rtcp-fb:96 goog-remb
a=rtcp-fb:96 transport-cc
a=rtpmap:98 VP9/90000
a=rtcp-fb:98 ccm fir
a=rtcp-fb:98 nack
a=rtcp-fb:98 nack pli
a=rtcp-fb:98 goog-remb
a=rtcp-fb:98 transport-cc
a=rtpmap:100 red/90000
a=rtpmap:127 ulpfec/90000
a=rtpmap:125 H264/90000
a=rtcp-fb:125 ccm fir
a=rtcp-fb:125 nack
a=rtcp-fb:125 nack pli
a=rtcp-fb:125 goog-remb
a=rtcp-fb:125 transport-cc
a=fmtp:125 level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=42e01f
a=rtpmap:97 rtx/90000
a=fmtp:97 apt=96
a=rtpmap:99 rtx/90000
a=fmtp:99 apt=98
a=rtpmap:101 rtx/90000
a=fmtp:101 apt=100
a=rtpmap:124 rtx/90000
a=fmtp:124 apt=125
a=ssrc-group:FID 3556063086 242494845
a=ssrc:3556063086 cname:duTyMEGqb0G2cnOB
a=ssrc:3556063086 msid:ARDAMS 101
a=ssrc:3556063086 mslabel:ARDAMS
a=ssrc:3556063086 label:101
a=ssrc:242494845 cname:duTyMEGqb0G2cnOB
a=ssrc:242494845 msid:ARDAMS 101
a=ssrc:242494845 mslabel:ARDAMS
a=ssrc:242494845 label:101
m=application 9 DTLS/SCTP 5000
c=IN IP4 0.0.0.0
a=ice-ufrag:Len2
a=ice-pwd:w0zCZt7ROLAngg0y/eFjD4Yj
a=ice-options:trickle renomination
a=fingerprint:sha-256 D9:45:AC:F2:1A:2D:2A:25:10:64:E1:28:D3:C9:33:D1:47:41:53:54:02:44:D2:45:4D:32:83:5E:AD:3E:AC:0D
a=setup:actpass
a=mid:data
a=sctpmap:5000 webrtc-datachannel 1024

----------
Answer from windows PC:
v=0
o=- 1191356116277972924 2 IN IP4 127.0.0.1
s=-
t=0 0
a=group:BUNDLE audio video data
a=msid-semantic: WMS local_av_stream
m=audio 9 UDP/TLS/RTP/SAVPF 111 103 104 9 102 0 8 106 105 13 110 112 113 126
c=IN IP4 0.0.0.0
a=rtcp:9 IN IP4 0.0.0.0
a=ice-ufrag:s+2M
a=ice-pwd:hSySAHZnd5RdG1dxtUiLUgsg
a=ice-options:trickle
a=fingerprint:sha-256 93:46:B8:DF:0D:51:6C:E3:68:9C:2E:34:B1:1D:B9:86:E6:88:BC:09:E5:7F:A0:0F:2C:35:E2:D5:6C:79:0A:24
a=setup:active
a=mid:audio
a=extmap:1 urn:ietf:params:rtp-hdrext:ssrc-audio-level
a=sendrecv
a=rtcp-mux
a=rtpmap:111 opus/48000/2
a=rtcp-fb:111 transport-cc
a=fmtp:111 minptime=10;useinbandfec=1
a=rtpmap:103 ISAC/16000
a=rtpmap:104 ISAC/32000
a=rtpmap:9 G722/8000
a=rtpmap:102 ILBC/8000
a=rtpmap:0 PCMU/8000
a=rtpmap:8 PCMA/8000
a=rtpmap:106 CN/32000
a=rtpmap:105 CN/16000
a=rtpmap:13 CN/8000
a=rtpmap:110 telephone-event/48000
a=rtpmap:112 telephone-event/32000
a=rtpmap:113 telephone-event/16000
a=rtpmap:126 telephone-event/8000
a=ssrc:4242883319 cname:+Il6btbpE6oZ4x7P
a=ssrc:4242883319 msid:local_av_stream local_audio
a=ssrc:4242883319 mslabel:local_av_stream
a=ssrc:4242883319 label:local_audio
m=video 9 UDP/TLS/RTP/SAVPF 96 98 100 127 97 99 101
c=IN IP4 0.0.0.0
a=rtcp:9 IN IP4 0.0.0.0
a=ice-ufrag:s+2M
a=ice-pwd:hSySAHZnd5RdG1dxtUiLUgsg
a=ice-options:trickle
a=fingerprint:sha-256 93:46:B8:DF:0D:51:6C:E3:68:9C:2E:34:B1:1D:B9:86:E6:88:BC:09:E5:7F:A0:0F:2C:35:E2:D5:6C:79:0A:24
a=setup:active
a=mid:video
a=extmap:2 urn:ietf:params:rtp-hdrext:toffset
a=extmap:3 http://www.webrtc.org/experiments/rtp-hdrext/abs-send-time
a=extmap:4 urn:3gpp:video-orientation
a=extmap:5 http://www.ietf.org/id/draft-holmer-rmcat-transport-wide-cc-extensions-01
a=extmap:6 http://www.webrtc.org/experiments/rtp-hdrext/playout-delay
a=sendrecv
a=rtcp-mux
a=rtcp-rsize
a=rtpmap:96 VP8/90000
a=rtcp-fb:96 goog-remb
a=rtcp-fb:96 transport-cc
a=rtcp-fb:96 ccm fir
a=rtcp-fb:96 nack
a=rtcp-fb:96 nack pli
a=rtpmap:98 VP9/90000
a=rtcp-fb:98 goog-remb
a=rtcp-fb:98 transport-cc
a=rtcp-fb:98 ccm fir
a=rtcp-fb:98 nack
a=rtcp-fb:98 nack pli
a=fmtp:98 x-google-profile-id=0
a=rtpmap:100 red/90000
a=rtpmap:127 ulpfec/90000
a=rtpmap:97 rtx/90000
a=fmtp:97 apt=96
a=rtpmap:99 rtx/90000
a=fmtp:99 apt=98
a=rtpmap:101 rtx/90000
a=fmtp:101 apt=100
a=ssrc-group:FID 818681678 2293923236
a=ssrc:818681678 cname:+Il6btbpE6oZ4x7P
a=ssrc:818681678 msid:local_av_stream local_video
a=ssrc:818681678 mslabel:local_av_stream
a=ssrc:818681678 label:local_video
a=ssrc:2293923236 cname:+Il6btbpE6oZ4x7P
a=ssrc:2293923236 msid:local_av_stream local_video
a=ssrc:2293923236 mslabel:local_av_stream
a=ssrc:2293923236 label:local_video
m=application 9 DTLS/SCTP 5000
c=IN IP4 0.0.0.0
b=AS:30
a=ice-ufrag:s+2M
a=ice-pwd:hSySAHZnd5RdG1dxtUiLUgsg
a=ice-options:trickle
a=fingerprint:sha-256 93:46:B8:DF:0D:51:6C:E3:68:9C:2E:34:B1:1D:B9:86:E6:88:BC:09:E5:7F:A0:0F:2C:35:E2:D5:6C:79:0A:24
a=setup:active
a=mid:data
a=sctpmap:5000 webrtc-datachannel 1024

**Comment by dsp85103** [comment]
The signaler was made by websocket myself. It can work correctly. 
I add local media stream of android to media track of peerconnection before send offer.
Android could successful to play from PC media stream (video&audio).  
PC couldn't render video frame from android.  But pc can play audio from android audio frame.

Thank u. 
I come from Taiwan. Maybe english not correct at all. ;) But I try hard to express all.

**Comment by djee-ms** [comment]
Ok so the offer and answer both contain some 2-way (`a=sendrecv`) tracks:
- audio track (`m=audio 9 UDP/TLS/RTP/SAVPF 111`) using the OPUS codec (`a=rtpmap:111 opus/48000/2`)
- video track (`m=video 9 UDP/TLS/RTP/SAVPF 96`) using the VP8 codec (`a=rtpmap:96 VP8/90000`)

The fact that the audio works means this is most likely a problem with the video codec or frame callbacks. Since VP8 seems to be selected and supported by both sides, and you tell me the TrackAdded callback is not invoked, I am tempted to say that an error occurs when receiving the answer. Unfortunately there is no easy way to debug that; this probably requires stepping into the C++ code to see why the answer is not applied correctly, and why it doesn't fire the TrackAdded callback. And the C++ NuGet packages do not have any PDB so this requires [recompiling the core dependencies](https://microsoft.github.io/MixedReality-WebRTC/manual/building-core.html), which is fairly complex.

At that moment we do not have official support for Android, so unfortunately I cannot really commit to doing that myself, as we're already pretty busy with other works.

Can I just confirm, on the PC side is this a C# project in Unity or a standalone C# .NET app?

**Comment by dsp85103** [comment]
OK. I realized. Very thank you reply me. This project just my test case. 
My pc side is a standalone C# .NET app.
Because I wanna make a windows-android live stream. But windows seems like not best one library for C# dotNET app.
So I tried a lot of library for windows. The MixedReality-WebRtc seems like full support in my case.
Never mind. I'll keep try another library in my case. Thank you for help. :)
